### PR TITLE
[Local Tensor] Replace dry_run.py with fake mode implementation

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #2057

Replaces `dry_run.py` implementation with fake PG mode for DRY_RUN configuration validation. This PR also adds support of Local tensor mode to provide deeper validation coverage.

**Note:** Currently returns early before `init_weights()` if using local tensor mode due to some limitation of local tensor, which will be fixed by https://github.com/pytorch/pytorch/pull/166540 .

### CODE REVIEW

File: torchtitan/train.py [hunk 1]
```diff
@@ -718,6 +723,13 @@ def main(trainer_class: type[Trainer]) -> None:
     try:
         trainer = trainer_class(config)
 
+        # TODO(local_tensor): Remove this special case once LocalTensor supports
+        # init_weights(). In local tensor mode, skip training/checkpointing as the
+        # model is not fully initialized
+        if config.comm.local_tensor_mode:
```
Comment (wconstab) [inline torchtitan/train.py]: probably a naive question. what's the advantage of localtensor over faketensor if we're just running the parallelization setup code?

localtensor would do the actual compute and give correct results, but it would run (more slowly) on a single gpu since it simulates each rank's operations.  fake tensor would skip all the compute and run (more quickly, i think) on a single CPU.

Do we intend to run with numerics or just smoke-test that we don't get API errors along the way to setup?

Comment (fegin) [reply torchtitan/train.py]: > Do we intend to run with numerics or just smoke-test that we don't get API errors along the way to setup?

For DryRun mode that validates the configurations, fake backend should be enough. But I also want to use this infra to enable some tests that do not require GPUs, running one step and verify the output. We are putting all the different parallelisms test into integration tests which require one H100 machine. The queuing time is going to become longer.

Also, DeviceMesh uses tensor operations, if we want to verify DeviceMesh operations on all ranks, we will need LocalTensor. Fake backend only allow you to verify rank0 DeviceMesh behavior, though this should not be a big deal as we mostly do SPMD.

Comment (fegin) [reply torchtitan/train.py]: I re-think you comment. I think I should not couple dry run mode with local tensor mode, which my intention was to use it for light-weight integration tests.

I add another option to enable ONLY fake PG mode. Local tensor mode depends on fake PG mode and dry run only requires fake PG mode.

Comment (wconstab) [reply torchtitan/train.py]: ok- that makes sense! i like the idea of using localtensor mode for running actual numerics validation. it just seemed overkill if used for dry-run.

File: torchtitan/config/job_config.py [hunk 2]
```diff
@@ -791,6 +791,15 @@ class Comm:
     save_traces_file_prefix: str = "rank_"
     """Flight recorder trace files prefix"""
 
+    fake_backend: bool = False
+    """Fake comm backend for dry run mode only"""
+
+    local_tensor_mode: bool = False
+    """
+    Local tensor mode, for debugging purposes. This is an experimental feature.
```
Comment (wwwjn) [inline torchtitan/config/job_config.py]: Can you provide more description here, or in the `debugging.md`? Is it running the model on a single GPU or CPU?

File: torchtitan/train.py (hunk 1 above)
Comment (wwwjn) [reply torchtitan/train.py]: I think the local tensor mode might help to some extent to debug numerics, however I met numeric issue because of DTensor, or missing communication. So we would recommend user debug with following order:  dry_run mode -> local tensor model -> parallelism , right?

Comment (fegin) [reply torchtitan/train.py]: IMO, dry run mode is purely debugging the setup of the trainer phase. The setup means the trainer configurations, DeviceMesh setting up and parallelism configurations. For an end user of TorchTitan, dry run mode is mostly useful to detect configurations error before launching a large scale training. For TorchTitan developers like us, dry run mode is useful as an early debugging signal when developing trainer, components and parallelisms (e.g., parallelize.py).

LocalTensor, on the other hand is useful to actually check what happen during the forward and backward. While fake tensor can also help, it doesn't actually incur the communication and computation which may hide the issues, and when the computation involve data dependent op, it will fail. For example, you won't be able to debug CP load balancing with fake tensor.

File: torchtitan/config/job_config.py (hunk 2 above)
Comment (fegin) [reply torchtitan/config/job_config.py]: Done

File: torchtitan/train.py [hunk 3]
```diff
@@ -208,6 +209,12 @@ def __init__(self, job_config: JobConfig):
             self.loss_fn, self.gradient_accumulation_steps
         )
 
+        # TODO(local_tensor): Remove this early return once LocalTensor supports
```
Comment (tianyu-l) [inline torchtitan/train.py]: https://github.com/pytorch/pytorch/pull/166540 is merged, shall we remove this early return?

File: torchtitan/train.py [hunk 4]
```diff
@@ -360,15 +367,19 @@ def __init__(self, job_config: JobConfig):
...
             base_folder=job_config.job.dump_folder,
         )
 
-        world_size = int(os.environ["WORLD_SIZE"])
-        parallelism_config = job_config.parallelism
+        if job_config.comm.local_tensor_mode:
+            if not job_config.comm.fake_backend:
+                raise ValueError("LocalTensor can only be used with fake backend.")
+            lm = _local_tensor.LocalTensorMode(world_size)
+            lm.__enter__()
```
Comment (tianyu-l) [inline torchtitan/train.py]: can we put this in `init_fake_mode`?

File: torchtitan/train.py [hunk 5]
```diff
@@ -718,6 +729,13 @@ def main(trainer_class: type[Trainer]) -> None:
     try:
         trainer = trainer_class(config)
 
+        # TODO(local_tensor): Remove this special case once LocalTensor supports
```
Comment (tianyu-l) [inline torchtitan/train.py:726]: similarly, can we remove this now?

File: torchtitan/distributed/utils.py [hunk 6]
```diff
@@ -258,12 +258,43 @@ def maybe_enable_amp(
...
+
+    Returns:
+        The world size
+    """
+    torch.distributed.init_process_group(
+        "fake",
+        rank=0,
+        world_size=world_size,
+    )
+    return world_size
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py]: `world_size` is already available as input, so we don't need to return. Whoever calls this method has this value.
Moreover I think this might be too simple to be a standalone function.

File: torchtitan/train.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/train.py]: I'm confused by the options here.

IIUC fake backend doesn't mean fake tensor right? What happens if `local_tensor_mode=False` but `fake_backend=True`?

Since this is user facing, I feel it might be more clear to organize options based on user intent. E.g. `comm_mode = "dry" / "local"` instead of providing multiple knobs which only functions when combined properly.

Comment (fegin) [reply torchtitan/train.py]: fake_backend=True, just means that we use fake backend for the communication. It doesn't have to be used with comm_mode if we don't care about accuracy. One computation will be done locally on rank0 but the collectives will not. For local_tensor_mode, all the computation (rank0 - rankN-1) will be done on rank0 and the collectives will be simulated as well.

I'm okay to combine the two together.

Comment (tianyu-l) [reply torchtitan/train.py]: With fake_backend=True, is rank0 the only device which participated in computation? Or does each rank compute their own stuff without actual comms?

Comment (fegin) [reply torchtitan/train.py]: @tianyu-l All ranks will perform the computation locally, but it actually doesn't matter because ranks don't talk to each others. For dry run mode, we always launch with one rank only (but fake it as there are N ranks).

File: torchtitan/train.py (hunk 3 above)
Comment (fegin) [reply torchtitan/train.py]: There are still some gaps. I updated the comment.

File: torchtitan/train.py (hunk 5 above)
Comment (fegin) [reply torchtitan/train.py:726]: There are still some gaps. I updated the comment.

**Comment by dolpm** [comment]
try https://github.com/pytorch/pytorch/pull/168310 to fix the seed issue

File: torchtitan/distributed/utils.py [hunk 7]
```diff
@@ -258,12 +258,56 @@ def maybe_enable_amp(
...
+    if comm_mode == "local_tensor":
+        from torch.distributed import _local_tensor
+
+        lm = _local_tensor.LocalTensorMode(world_size)
+        lm.__enter__()
+
+        # TODO: remove this once the root cause is figured out
+        torch.manual_seed(42)
+
+    return world_size
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py]: no need to `return world_size` any more?

File: torchtitan/config/job_config.py [hunk 8]
```diff
@@ -791,6 +791,22 @@ class Comm:
     save_traces_file_prefix: str = "rank_"
     """Flight recorder trace files prefix"""
 
+    comm_mode: Literal["default", "fake_backend", "local_tensor"] = "default"
```
Comment (tianyu-l) [inline torchtitan/config/job_config.py]: nit: `comm.comm_mode` sounds redundant, I would propose
```suggestion
    backend: Literal["default", "fake", "local"] = "default"
```
WDYT?

Comment (fegin) [reply torchtitan/config/job_config.py]: I will change it to `mode`. The reason why I didn't pick backend is that `local_tensor` is technically not a PG backend. It utilizes the PG fake backend. So using backend with `local` and `fake` are not very accurate.

**Comment by fegin** [comment]
We removes all the early return and hacks for local tensor except for the one to prevent from actual training. The reason why we cannot do an actual training is that LocalTensor doesn't support resize. This PR is ready to review and land.

**Comment by wconstab** [comment]
Fake backend takes the path of one particular rank, which is configured when initializing fake pg. commonly rank 0 is used. This means if rank 0 exits early then the whole fake simulation exits early and the other paths aren't tested.

Localtensor actually runs the computation for every rank, but it is only designed to support truly spmd code. I'm not sure what it does if some ranks branch out. Cc @dzmitry-huba

**Comment by dzmitry-huba** [comment]
Temporary deviations while executing under Local Tensor are possible with special annotations for the code regions representing deviation. However overall execution across ranks must follow an SPMD.

**Comment by fegin** [comment]
Just want to add some information on top of the above explanation. There is also FakeTensorMode. So technically, we can use FakeTensorMode + FakeBackend. But the if/else case is not supported by FakeTensorMode, which makes sense since the content is fake. As a result, I decided to not enable FakeTensorMode and to use FakeBackend only. So we can verify the if/else with fake_backend but with some constraints @wconstab mentioned.
